File: Bukkit/src/main/java/me/mrkirby153/kcutils/paginate/Paginator.java
package me.mrkirby153.kcutils.paginate;

import java.util.List;

/**
 * A paginator utility. Used to break a large list into multiple sub-lists
 *
 * @param <T>
 */
public class Paginator<T> {

    private final List<T> items;

    private final int itemsPerPage;

    public Paginator(List<T> items, int itemsPerPage) {
        this.items = items;
        this.itemsPerPage = itemsPerPage;
    }

    /**
     * Gets the amount of pages in the paginator
     *
     * @return The maximum number of pages
     */
    public int getMaxPages() {
        return (int) Math.ceil((double) this.items.size() / (double) this.itemsPerPage);
    }

    /**
     * Gets a list of items in a page of the paginator
     *
     * @param page The page
     * @return A list of items
     */
    public List<T> getPage(int page) {
        int startIndex = this.itemsPerPage * (page - 1);
        int endIndex = Math.min(this.itemsPerPage * page, items.size());
        return this.items.subList(startIndex, endIndex);
    }
}

File: Common/src/main/kotlin/me/mrkirby153/kcutils/paginate/Paginator.kt
package me.mrkirby153.kcutils.paginate

/**
 * A paginator utility. Used to break a large list into multiple sub-lists
 *
 * @param T The type of the paginator
 */
class Paginator<out T>(private val items: List<T>, private val itemsPerPage: Int) {

    /**
     * Gets the amount of pages in the paginator
     *
     * @return The maximum number of pages
     */
    val maxPages: Int
        get() = Math.ceil(this.items.size.toDouble() / this.itemsPerPage.toDouble()).toInt()

    /**
     * Gets a list of items in a page of the paginator
     *
     * @param page The page
     * @return A list of items
     */
    fun getPage(page: Int): List<T> {
        val startIndex = this.itemsPerPage * (page - 1)
        val endIndex = Math.min(this.itemsPerPage * page, items.size)
        return this.items.subList(startIndex, endIndex)
    }
}
